src/frameobject.ts

```diff
@@ -141,6 +143,36 @@ class Py_FrameObject implements IPy_FrameObj {
             }
         }
     }
+
+    emptyStack() {
+        while (this.stack.length > 0) {
+            this.pop();
+        }
+    }
+    raise_exception_here(t: Thread, message: string, type: string): void {
+        var val: IPy_Object = (<any> builtins)[type];
+        t.exc = val;
+        this.push(val);
+        t.addToTraceback(message);
+        this.tryCatchException(t);
+    }
+
+    tryCatchException(t: Thread): boolean {
+        // Whenever an exception occurs, tries to find a handler and if it can't outputs the traceback 
+        // search in current frame
+        if (!find_exception_handler(this)) {
+            // Search for exception handler in previous frames
```

Comment:
The frame shouldn't be searching previous frames. The thread should be doing this.

Think of it this way: The frame object should *only* be concerned about *itself* -- not other frames. The thread object should be concerned about the set of frames that are on its callstack.

The frame is responsible for handling return values passed to it, and  for trying to handle exceptions within itself. The thread object is responsible for providing the return values to the frame, and for letting the frame know that an uncaught exception is passing through -- and asking it if its method can handle the exception or not. However, the thread has *no idea* what the frame is doing to handle the exception.

Keeping abstractions clean makes the implementation cleaner, and lets us reuse Thread logic for both the JVM and Python.